File: scripts/check-links.py
#!/usr/bin/env python3
"""sapwood pre-publish dead-link checker (#340).

Invocation (from the repository root):

    python3 scripts/check-links.py

Optionally check a different checkout (e.g. a worktree or an older commit's
tree) by passing its path:

    python3 scripts/check-links.py /path/to/other/checkout

Pass `--internal-only` to skip the external-link section entirely (no
network, no `gh` call) and check only relative links/anchors. This is the
mode wired into CI, where a `gh`-authenticated, networked check would be
flaky and would gate merges on outside URLs sapwood does not control:

    python3 scripts/check-links.py --internal-only

Stdlib only -- no new dependency, nothing to `npm install`.

Scope: README.md + docs/ entry pages -- every top-level `docs/*.md`, plus
`docs/dev-guide/README.md` -- and everything those pages link to, followed
transitively over relative Markdown links. This is the set #340's
acceptance criteria call "README + docs entry pages."

What it checks:
  - Every relative Markdown link (`[text](path)` or `[text](path#anchor)`)
    resolves to a file that exists.
  - Every `#anchor` (same-file or cross-file) resolves to a real heading in
    the target file, using GitHub's own heading-slug algorithm: lowercase,
    strip Markdown emphasis/code markers, drop every character that is not
    a Unicode letter/mark/decimal-digit/hyphen/space (this deliberately
    excludes Python's `\\w`, which over-keeps symbol categories like circled
    digits that GitHub's slugger does not), map each remaining literal
    space to its own hyphen (an em dash sitting between two spaces produces
    a double hyphen, because the dash itself is dropped but both spaces
    survive -- GitHub does not collapse runs), and de-duplicate repeated
    slugs the way GitHub does: the second `## Foo` on a page gets `#foo-1`,
    the third `#foo-2`, and so on.
  - Every `https://github.com/herehigher/sapwood/issues/N` or `/pull/N`
    link is checked with `gh api` rather than a raw HTTP request: this repo
    is currently **private**, so an unauthenticated HTTP probe 404s on
    every such URL regardless of whether the issue/PR actually exists.
    Other `github.com/herehigher/sapwood/...` links (workflow badges, the
    private-vulnerability-reporting URL, etc.) can't be probed the same way
    and are reported as `UNVERIFIED-PRIVATE` rather than silently passed.
  - Every other `http(s)` external link gets a best-effort HEAD (falling
    back to GET on 403/405) request.

Exit code: non-zero iff at least one internal link/anchor problem was
found. External-link results are informational (network- and
private-repo-dependent) and never affect the exit code.
"""
import os
import re
import subprocess
import sys
import unicodedata
import urllib.error
import urllib.request

REPO = "herehigher/sapwood"

LINK_RE = re.compile(r'\[([^\]]*)\]\(([^)]+)\)')


def slugify(text):
    """Reproduce GitHub's heading-slug algorithm (github-slugger behavior)."""
    text = re.sub(r'[`*_]', '', text)  # strip inline code/emphasis markers
    text = text.lower()
    kept = []
    for ch in text:
        if ch == ' ' or ch == '-':
            kept.append(ch)
        else:
            cat = unicodedata.category(ch)
            # Keep letters (L*), combining marks (M*), and decimal digits
            # (Nd) only -- NOT Python's \w, which also keeps Unicode
            # category No/Nl (e.g. circled digits like the gate marks used
            # in this repo's own headings) that GitHub's slugger strips.
            if cat[0] == 'L' or cat[0] == 'M' or cat == 'Nd':
                kept.append(ch)
    text = ''.join(kept).strip()
    # Map each literal space to its own hyphen -- do NOT collapse runs, so
    # "table — reading" (em dash between two spaces, dash itself dropped
    # above) yields "table--reading", matching GitHub exactly.
    return text.replace(' ', '-')


def headings_of(path):
    """Return {slug: True} for every heading in a Markdown file, with
    GitHub-style duplicate-slug suffixing (-1, -2, ...)."""
    slugs = {}
    try:
        with open(path, encoding='utf-8') as f:
            content = f.read()
    except (FileNotFoundError, UnicodeDecodeError):
        return slugs
    for line in content.splitlines():
        m = re.match(r'^(#{1,6})\s+(.*)$', line)
        if not m:
            continue
        base = slugify(m.group(2))
        slug = base
        i = 1
        while slug in slugs:
            slug = f"{base}-{i}"
            i += 1
        slugs[slug] = True
    return slugs


def collect_scope(root):
    """Seed on README + top-level docs entry pages, then follow relative
    .md links transitively."""
    seeds = [
        "README.md",
        "CONTRIBUTING.md",
        "SECURITY.md",
        "docs/PLAN.md",
        "docs/getting-started.md",
        "docs/configuration.md",
        "docs/security.md",
        "docs/role-paradigm.md",
        "docs/troubleshooting.md",
        "docs/supervision.md",
        "docs/dev-guide/README.md",
    ]
    scope = set()
    queue = [f for f in seeds if os.path.exists(os.path.join(root, f))]
    while queue:
        rel = queue.pop()
        if rel in scope:
            continue
        scope.add(rel)
        full = os.path.join(root, rel)
        if not os.path.exists(full):
            continue
        with open(full, encoding='utf-8') as f:
            content = f.read()
        for _, target in LINK_RE.findall(content):
            target = target.split(' ')[0].strip('<>')
            if target.startswith(('http://', 'https://', 'mailto:', '#')):
                continue
            path_part = target.split('#')[0]
            if not path_part:
                continue
            newpath = os.path.normpath(os.path.join(os.path.dirname(rel), path_part))
            if newpath.endswith('.md') and os.path.exists(os.path.join(root, newpath)) and newpath not in scope:
                queue.append(newpath)
    return scope


def check_file(root, rel, results, external_links):
    full = os.path.join(root, rel)
    with open(full, encoding='utf-8') as f:
        content = f.read()
    for _, target in LINK_RE.findall(content):
        raw_target = target
        target = target.split(' ')[0].strip('<>')
        if not target:
            continue
        if target.startswith(('http://', 'https://')):
            external_links.setdefault(target, []).append(rel)
            continue
        if target.startswith('mailto:'):
            continue
        if target.startswith('#'):
            anchor = target[1:]
            if anchor and anchor not in headings_of(full):
                results.append((rel, raw_target, 'BROKEN ANCHOR (same-file)', f'no heading slug "{anchor}" in {rel}'))
            continue
        path_part, _, anchor = target.partition('#')
        newpath = os.path.normpath(os.path.join(os.path.dirname(rel), path_part))
        newfull = os.path.join(root, newpath)
        if not os.path.exists(newfull):
            results.append((rel, raw_target, 'BROKEN LINK', f'target does not exist: {newpath}'))
            continue
        if anchor and newpath.endswith('.md') and anchor not in headings_of(newfull):
            results.append((rel, raw_target, 'BROKEN ANCHOR', f'no heading slug "{anchor}" in {newpath}'))


def check_external(external_links):
    ok, fail, unverified = 0, 0, 0
    for url in sorted(external_links):
        srcs = ', '.join(sorted(set(external_links[url])))
        m = re.match(rf'https://github\.com/{re.escape(REPO)}/(issues|pull)/(\d+)', url)
        if m:
            kind = 'issues' if m.group(1) == 'issues' else 'pulls'
            r = subprocess.run(
                ['gh', 'api', f'repos/{REPO}/{kind}/{m.group(2)}', '--jq', '.state'],
                capture_output=True, text=True)
            if r.returncode == 0:
                ok += 1
                print(f"  [OK (gh api: {r.stdout.strip()})] {url}  (linked from: {srcs})")
            else:
                fail += 1
                print(f"  [ERROR (gh api: {r.stderr.strip()})] {url}  (linked from: {srcs})")
            continue
        if url.startswith(f'https://github.com/{REPO}/'):
            # Repo-relative link that isn't an issue/PR (workflow badge,
            # advisories-new form, etc.) -- while the repo is private, an
            # unauthenticated probe 404s regardless of validity, so this is
            # reported rather than silently assumed OK.
            unverified += 1
            print(f"  [UNVERIFIED-PRIVATE] {url}  (linked from: {srcs})")
            continue
        req = urllib.request.Request(url, method='HEAD', headers={'User-Agent': 'sapwood-check-links/1.0'})
        try:
            with urllib.request.urlopen(req, timeout=8) as resp:
                status = 'OK' if resp.status < 400 else f'HTTP {resp.status}'
        except urllib.error.HTTPError as e:
            if e.code in (403, 405):
                try:
                    req2 = urllib.request.Request(url, method='GET', headers={'User-Agent': 'sapwood-check-links/1.0'})
                    with urllib.request.urlopen(req2, timeout=8) as resp2:
                        status = 'OK' if resp2.status < 400 else f'HTTP {resp2.status}'
                except Exception as e2:
                    status = f'ERROR {e2}'
            else:
                status = f'HTTP {e.code}'
        except Exception as e:
            status = f'ERROR {e}'
        if status == 'OK':
            ok += 1
        elif status.startswith('ERROR'):
            unverified += 1
        else:
            fail += 1
        print(f"  [{status}] {url}  (linked from: {srcs})")
    return ok, fail, unverified


def main():
    args = sys.argv[1:]
    internal_only = '--internal-only' in args
    positional = [a for a in args if a != '--internal-only']
    root = os.path.abspath(positional[0]) if positional else os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    scope = sorted(collect_scope(root))
    print(f"Scope ({len(scope)} files, root={root}):")
    for f in scope:
        print(f"  - {f}")
    print()

    results = []
    external_links = {}
    for rel in scope:
        check_file(root, rel, results, external_links)

    print(f"=== Internal link/anchor check: {len(results)} problem(s) ===")
    for rel, target, kind, detail in results:
        print(f"  [{kind}] {rel} -> {target}  ({detail})")
    if not results:
        print("  (none)")

    if internal_only:
        return 1 if results else 0

    print()
    print(f"=== External links found ({len(external_links)} unique) ===")
    print(f"(repo is private -> unauthenticated HTTP to github.com/{REPO}/*")
    print(" URLs 404s regardless of validity; issue/PR links are checked via `gh api` instead)")
    ok, fail, unverified = check_external(external_links)
    print()
    print(f"External summary: {ok} ok, {fail} non-2xx/3xx/gh-api-error, {unverified} unverified/unreachable")

    return 1 if results else 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/check-links.ts
// sapwood dead-link checker — `npm run check-links` → `node scripts/check-links.ts --internal-only`.
// Node built-ins only (node:fs/path/child_process + fetch), no dependency. Plain Node 24
// type-stripped TypeScript, same pattern as release.ts: erasable syntax only, local imports
// carry an explicit `.ts` extension.
//
// Optionally check a different checkout (e.g. a worktree or an older commit's tree) by
// passing its path: `node scripts/check-links.ts /path/to/other/checkout`.
//
// `--internal-only` skips the external-link section entirely (no network, no `gh` call) and
// checks only relative links/anchors. This is the mode wired into CI, where a
// `gh`-authenticated, networked check would be flaky and would gate merges on outside URLs
// sapwood does not control.
//
// Scope: README.md + docs/ entry pages (see SEED_PAGES below) — and everything those pages
// link to, followed transitively over relative Markdown links. A seed that does not exist is
// reported as a MISSING SEED problem rather than silently dropped, so a doc move that forgets
// to update this list fails loudly instead of shrinking the checked scope unnoticed.
//
// What it checks:
//   - Every relative Markdown link (`[text](path)` or `[text](path#anchor)`) resolves to a
//     file that exists.
//   - Every `#anchor` (same-file or cross-file) resolves to a real heading in the target
//     file, using GitHub's own heading-slug algorithm (see engine/src/util/markdown-slug.ts).
//   - Every `https://github.com/herehigher/sapwood/issues/N` or `/pull/N` link is checked
//     with `gh api` rather than a raw HTTP request: this repo is currently **private**, so an
//     unauthenticated HTTP probe 404s on every such URL regardless of whether the issue/PR
//     actually exists. Other `github.com/herehigher/sapwood/...` links (workflow badges, the
//     private-vulnerability-reporting URL, etc.) can't be probed the same way and are
//     reported as `UNVERIFIED-PRIVATE` rather than silently passed.
//   - Every other `http(s)` external link gets a best-effort HEAD (falling back to GET on
//     403/405) request.
//
// Exit code: non-zero iff at least one internal link/anchor problem was found. External-link
// results are informational (network- and private-repo-dependent) and never affect the exit
// code.
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, realpathSync } from "node:fs";
import { dirname, join, normalize, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { headingSlugs } from "../engine/src/util/markdown-slug.ts";

const REPO = "herehigher/sapwood";

const LINK_RE = /\[([^\]]*)\]\(([^)]+)\)/g;

function headingsOf(path: string): Set<string> {
  try {
    return headingSlugs(readFileSync(path, "utf8"));
  } catch {
    return new Set();
  }
}

function stripAngleBrackets(s: string): string {
  return s.replace(/^[<>]+/, "").replace(/[<>]+$/, "");
}

const SEED_PAGES = [
  "README.md",
  "CONTRIBUTING.md",
  "SECURITY.md",
  "docs/README.md",
  "docs/PLAN.md",
  "docs/guide/getting-started.md",
  "docs/guide/configuration.md",
  "docs/security.md",
  "docs/reference/role-paradigm.md",
  "docs/guide/troubleshooting.md",
  "docs/guide/supervision.md",
  "docs/dev-guide/README.md",
];

// Seeds on README + top-level docs entry pages, then follows relative .md links
// transitively. Traversal order doesn't matter — the result is a set, sorted by the caller.
// `missingSeeds` (seeds absent from disk) is returned separately rather than just skipped, so
// the caller can surface it as a problem instead of quietly shrinking the checked scope.
function collectScope(root: string): { scope: Set<string>; missingSeeds: string[] } {
  const scope = new Set<string>();
  const missingSeeds = SEED_PAGES.filter((f) => !existsSync(join(root, f)));
  const queue = SEED_PAGES.filter((f) => !missingSeeds.includes(f));
  while (queue.length > 0) {
    const rel = queue.pop();
    if (rel === undefined || scope.has(rel)) continue;
    scope.add(rel);
    const full = join(root, rel);
    if (!existsSync(full)) continue;
    const content = readFileSync(full, "utf8");
    for (const match of content.matchAll(LINK_RE)) {
      let target = (match[2] ?? "").split(" ")[0] ?? "";
      target = stripAngleBrackets(target);
      if (target.startsWith("http://") || target.startsWith("https://") || target.startsWith("mailto:") || target.startsWith("#")) continue;
      const pathPart = target.split("#")[0] ?? "";
      if (!pathPart) continue;
      const newpath = normalize(join(dirname(rel), pathPart));
      if (newpath.endsWith(".md") && existsSync(join(root, newpath)) && !scope.has(newpath)) queue.push(newpath);
    }
  }
  return { scope, missingSeeds };
}

type Problem = [rel: string, target: string, kind: string, detail: string];

function checkFile(root: string, rel: string, results: Problem[], externalLinks: Map<string, string[]>): void {
  const full = join(root, rel);
  const content = readFileSync(full, "utf8");
  for (const match of content.matchAll(LINK_RE)) {
    const rawTarget = match[2] ?? "";
    const target = stripAngleBrackets(rawTarget.split(" ")[0] ?? "");
    if (!target) continue;
    if (target.startsWith("http://") || target.startsWith("https://")) {
      const srcs = externalLinks.get(target) ?? [];
      srcs.push(rel);
      externalLinks.set(target, srcs);
      continue;
    }
    if (target.startsWith("mailto:")) continue;
    if (target.startsWith("#")) {
      const anchor = target.slice(1);
      if (anchor && !headingsOf(full).has(anchor)) {
        results.push([rel, rawTarget, "BROKEN ANCHOR (same-file)", `no heading slug "${anchor}" in ${rel}`]);
      }
      continue;
    }
    const hashIdx = target.indexOf("#");
    const pathPart = hashIdx === -1 ? target : target.slice(0, hashIdx);
    const anchor = hashIdx === -1 ? "" : target.slice(hashIdx + 1);
    const newpath = normalize(join(dirname(rel), pathPart));
    const newfull = join(root, newpath);
    if (!existsSync(newfull)) {
      results.push([rel, rawTarget, "BROKEN LINK", `target does not exist: ${newpath}`]);
      continue;
    }
    if (anchor && newpath.endsWith(".md") && !headingsOf(newfull).has(anchor)) {
      results.push([rel, rawTarget, "BROKEN ANCHOR", `no heading slug "${anchor}" in ${newpath}`]);
    }
  }
}

const ISSUE_PR_RE = new RegExp(`^https://github\\.com/${REPO}/(issues|pull)/(\\d+)`);

function isExecError(e: unknown): e is { stderr?: string } {
  return typeof e === "object" && e !== null && "stderr" in e;
}

async function probeHttp(url: string): Promise<string> {
  const headers = { "User-Agent": "sapwood-check-links/1.0" };
  try {
    const resp = await fetch(url, { method: "HEAD", headers, signal: AbortSignal.timeout(8000) });
    if (resp.status !== 403 && resp.status !== 405) return resp.status < 400 ? "OK" : `HTTP ${resp.status}`;
    // GitHub and some other hosts reject HEAD on certain routes — retry with GET, same as
    // the Python original, rather than reporting a false failure.
    try {
      const resp2 = await fetch(url, { method: "GET", headers, signal: AbortSignal.timeout(8000) });
      return resp2.status < 400 ? "OK" : `HTTP ${resp2.status}`;
    } catch (e2) {
      return `ERROR ${e2 instanceof Error ? e2.message : String(e2)}`;
    }
  } catch (e) {
    return `ERROR ${e instanceof Error ? e.message : String(e)}`;
  }
}

async function checkExternal(externalLinks: Map<string, string[]>): Promise<{ ok: number; fail: number; unverified: number }> {
  let ok = 0;
  let fail = 0;
  let unverified = 0;
  for (const url of [...externalLinks.keys()].sort()) {
    const srcs = [...new Set(externalLinks.get(url) ?? [])].sort().join(", ");
    const m = ISSUE_PR_RE.exec(url);
    if (m) {
      const kind = m[1] === "issues" ? "issues" : "pulls";
      try {
        // `gh api` rather than a raw HTTP request — see the module comment on why an
        // unauthenticated probe can't distinguish "private repo" from "does not exist".
        const state = execFileSync("gh", ["api", `repos/${REPO}/${kind}/${m[2]}`, "--jq", ".state"], { encoding: "utf8" }).trim();
        ok++;
        console.log(`  [OK (gh api: ${state})] ${url}  (linked from: ${srcs})`);
      } catch (e) {
        fail++;
        const detail = isExecError(e) && e.stderr ? e.stderr.trim() : e instanceof Error ? e.message : String(e);
        console.log(`  [ERROR (gh api: ${detail})] ${url}  (linked from: ${srcs})`);
      }
      continue;
    }
    if (url.startsWith(`https://github.com/${REPO}/`)) {
      unverified++;
      console.log(`  [UNVERIFIED-PRIVATE] ${url}  (linked from: ${srcs})`);
      continue;
    }
    const status = await probeHttp(url);
    if (status === "OK") ok++;
    else if (status.startsWith("ERROR")) unverified++;
    else fail++;
    console.log(`  [${status}] ${url}  (linked from: ${srcs})`);
  }
  return { ok, fail, unverified };
}

function repoRootFromThisFile(): string {
  return dirname(dirname(fileURLToPath(import.meta.url)));
}

async function main(argv: string[]): Promise<number> {
  const args = argv.slice(2);
  const internalOnly = args.includes("--internal-only");
  const positional = args.filter((a) => a !== "--internal-only");
  const root = positional.length > 0 && positional[0] ? resolve(positional[0]) : repoRootFromThisFile();

  const { scope: scopeSet, missingSeeds } = collectScope(root);
  const scope = [...scopeSet].sort();
  console.log(`Scope (${scope.length} files, root=${root}):`);
  for (const f of scope) console.log(`  - ${f}`);
  console.log();

  const results: Problem[] = [];
  for (const f of missingSeeds) results.push([f, "(seed)", "MISSING SEED", `seed file does not exist: ${f}`]);

  const externalLinks = new Map<string, string[]>();
  for (const rel of scope) checkFile(root, rel, results, externalLinks);

  console.log(`=== Internal link/anchor check: ${results.length} problem(s) ===`);
  for (const [rel, target, kind, detail] of results) console.log(`  [${kind}] ${rel} -> ${target}  (${detail})`);
  if (results.length === 0) console.log("  (none)");

  if (internalOnly) return results.length > 0 ? 1 : 0;

  console.log();
  console.log(`=== External links found (${externalLinks.size} unique) ===`);
  console.log(`(repo is private -> unauthenticated HTTP to github.com/${REPO}/*`);
  console.log(" URLs 404s regardless of validity; issue/PR links are checked via `gh api` instead)");
  const { ok, fail, unverified } = await checkExternal(externalLinks);
  console.log();
  console.log(`External summary: ${ok} ok, ${fail} non-2xx/3xx/gh-api-error, ${unverified} unverified/unreachable`);

  return results.length > 0 ? 1 : 0;
}

function isMain(): boolean {
  const entry = process.argv[1];
  if (!entry) return false;
  try {
    return realpathSync(entry) === realpathSync(fileURLToPath(import.meta.url));
  } catch {
    return false;
  }
}

if (isMain()) {
  main(process.argv).then((code) => process.exit(code));
}
